Remove static variables in bin_z64.c

ahmed-kamal2004 opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository.
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style).
- [ ] I've documented every `RZ_API` function and struct this PR changes.
- [ ] I've added tests that prove my changes are effective (required for changes to `RZ_API`).
- [ ] I've updated the [Rizin book](https://github.com/rizinorg/book) with the relevant information (if needed).

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

...

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->
Remove static var `n64_header` from file `bin_z64.c`, relying on `bin_obj` in RzBinObject to hold the state
...

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

...

ahmed-kamal2004 committed:
Remove static variables in bin_z64.c

Rot127 commented on librz/bin/p/bin_z64.c:
```diff
@@ -80,13 +79,16 @@ static bool check_buffer(RzBuffer *b) {
+		return false;
 	}
-	return false;
+	ut8 buf[sizeof(N64Header)] = { 0 };
+	rz_buf_read_at(b, 0, buf, sizeof(buf));
+	memcpy(hdr, buf, sizeof(N64Header));
```
You can directly read into the header, no need to first create a buffer on the stack and then copy it. You just need to cast the pointer.

ahmed-kamal2004 replied:
yeah, was following the old code

wargio replied:
As mentioned before the old code is wrong and bad. According to the old code the whole binary is big endian, so you should parse the header like the other plug-ins. I shared before some links to the nin3ds parser which uses the offset api of RzBuffer. Please do the same here.

ahmed-kamal2004 committed:
Refactor code

ahmed-kamal2004 commented:
@wargio @Rot127 
when I implemented as BE, is doesn't pass tests, I assume tests were written for LE, so the old code was LE
```

└─(01:04:01 on bin_z64 ✹)──> rz-test ./db/formats/z64                                 ──(Sat,Jun28)─┘
Running rz-test on linux-x64
Running from /home/ahmed-kamal/Every/gs/rizin/test
Loaded 3 tests.
[XX] 169 ms   /home/ahmed-kamal/Every/gs/rizin/test/db/formats/z64 z64: sections
RZ_NOPLUGINS=1 /usr/local/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -esearch.show_progress=false -eflirt.sigdb.load.home=false -N -qc iS bins/z64/demo.z64
-- stdout
--- expected
+++ actual
@@ -0,4 +0,4 @@
      paddr     size      vaddr    vsize align perm name type flags 
 -------------------------------------------------------------------
-0x00001000 0x13f000 0x80001000 0x13f000   0x0 -r-x text      
+0x00001000 0x13f000 0x00100080 0x13f000   0x0 -r-x text      

[XX] 170 ms   /home/ahmed-kamal/Every/gs/rizin/test/db/formats/z64 z64: maps
RZ_NOPLUGINS=1 /usr/local/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -esearch.show_progress=false -eflirt.sigdb.load.home=false -N -qc oml bins/z64/demo.z64
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
- 1 fd: 3 +0x00001000 0x80001000 * 0x8013ffff r-x fmap.text
+ 1 fd: 3 +0x00001000 0x00100080 * 0x0023f07f r-x fmap.text

[3/3]                       1 OK         0 BR        2 XX        0 FX
```

under that I implemented as LE until you see it

ahmed-kamal2004 committed:
use LE buffer reading in z64

Rot127 commented:
In this reddit post they say the endianess depends on the file name: https://www.reddit.com/r/emulation/comments/7hrvzp/the_three_different_n64_rom_formats_explained_for/

You would need to check the file name suffix and then load the header information accordingly.

Rot127 commented on librz/bin/p/bin_z64.c:
```diff
@@ -79,14 +78,36 @@ static bool check_buffer(RzBuffer *b) {
+		rz_buf_read_ble8_offset(buf, &offset, &hdr->x2, false) &&
+		rz_buf_read_ble8_offset(buf, &offset, &hdr->x3, false) &&
+		rz_buf_read_ble8_offset(buf, &offset, &hdr->x4, false) &&
+		rz_buf_read_le32_offset(buf, &offset, &hdr->ClockRate) &&
+		rz_buf_read_le32_offset(buf, &offset, &hdr->BootAddress) &&
+		rz_buf_read_le32_offset(buf, &offset, &hdr->Release) &&
+		rz_buf_read_le32_offset(buf, &offset, &hdr->CRC1) &&
+		rz_buf_read_le32_offset(buf, &offset, &hdr->CRC2) &&
+		rz_buf_read_le64_offset(buf, &offset, &hdr->UNK1) &&
+		rz_buf_read_offset(buf, &offset, (ut8 *)hdr->Name, sizeof(hdr->Name)) &&
+		rz_buf_read_le32_offset(buf, &offset, &hdr->UNK2) &&
+		rz_buf_read_le16_offset(buf, &offset, &hdr->UNK3) &&
+		rz_buf_read_ble8_offset(buf, &offset, &hdr->UNK4, false) &&
+		rz_buf_read_ble8_offset(buf, &offset, &hdr->ManufacturerID, false) &&
+		rz_buf_read_le16_offset(buf, &offset, &hdr->CartridgeID) &&
+		rz_buf_read_ble8_offset(buf, &offset, (ut8 *)&hdr->CountryCode, false) &&
+		rz_buf_read_ble8_offset(buf, &offset, &hdr->UNK5, false);
```
You intermix little and big endian reading here. You need to just use the `ble` version of the functinos

ahmed-kamal2004 replied:
I don't think I intermix them, I implemented to be as little indian, so I used `rz_buf_read_le32_offset` function, and set flag to `false`, please correct me if I misunderstood a function or something, what I am not sure about is `rz_buf_read_offset`
also, I will be going to change big indian

Rot127 replied:
Nah, you need to either read big or little endian depending to the file ending. See the link above.

wargio commented on librz/bin/p/bin_z64.c:
```diff
@@ -79,14 +78,36 @@ static bool check_buffer(RzBuffer *b) {
 	return !memcmp(magic, "\x80\x37\x12\x40", 4);
 }
 
+static bool n64_read_firm_hdr(RzBuffer *buf, N64Header *hdr) {
```
Rename this also from `n64_read_firm_hdr` to `n64_read_rom_hdr` since this is not a firmware.

ahmed-kamal2004 committed:
Add Little indian format and change tests

wargio commented on librz/bin/p/bin_z64.c:
```diff
@@ -75,34 +79,60 @@ static bool check_buffer(RzBuffer *b) {
+	}
+	(void)rz_buf_read_at(buf, 0, magic, sizeof(magic));
+	if (!memcmp(magic, "\x80\x37\x12\x40", 4)) {
+		return BIG_INDIAN_BYTE_ORDER;
```
ENDIAN, not INDIAN

wargio reviewed (changes requested):
^

wargio commented on librz/bin/p/bin_z64.c:
```diff
@@ -64,10 +68,9 @@ typedef struct {
 
-static N64Header n64_header;
-
 static ut64 baddr(RzBinFile *bf) {
-	return (ut64)rz_read_be32(&n64_header.BootAddress);
+	N64Header *hdr = bf->o->bin_obj;
+	return (ut64)rz_read_be32(&(hdr->BootAddress));
```
if the header is read correctly, then this is not needed and instead it should just return this:

```c
N64Header *hdr = bf->o->bin_obj;
return hdr->BootAddress;
```

wargio commented on librz/bin/p/bin_z64.c:
```diff
@@ -76,17 +79,65 @@ static bool check_buffer(RzBuffer *b) {
+}
+
+static int get_byte_order_format(RzBuffer *buf) {
+	ut8 magic[4];
```
```suggestion
	ut8 magic[4] = { 0 };
```

wargio commented on librz/bin/p/bin_z64.c:
```diff
@@ -76,17 +79,65 @@ static bool check_buffer(RzBuffer *b) {
+static bool n64_read_orm_hdr(RzBuffer *buf, N64Header *hdr) {
+	ut64 offset = 0;
+	size_t byte_order = get_byte_order_format(buf);
+	bool byte_order_flag = true;
```
```suggestion
	bool big_endian = true;
```

wargio commented on librz/bin/p/bin_z64.c:
```diff
@@ -76,17 +79,65 @@ static bool check_buffer(RzBuffer *b) {
+	size_t byte_order = get_byte_order_format(buf);
+	bool byte_order_flag = true;
+
+	switch (byte_order) {
+	case LITTLE_INDIAN_BYTE_ORDER:
+		byte_order_flag = false;
+		break;
+	case BIG_INDIAN_BYTE_ORDER:
+	default:
+		byte_order_flag = true;
+		break;
+	}
```
```suggestion
	switch (byte_order) {
	case LITTLE_ENDIAN_BYTE_ORDER:
		big_endian = false;
		break;
	case BIG_ENDIAN_BYTE_ORDER:
		big_endian = true;
		break;
	default:
		return false; // always fail if is not BE or LE
	}
```

wargio commented on librz/bin/p/bin_z64.c:
```diff
@@ -76,17 +79,65 @@ static bool check_buffer(RzBuffer *b) {
-		obj->bin_obj = memcpy(&n64_header, buf, sizeof(N64Header));
-		return true;
+	N64Header *hdr = RZ_NEW0(N64Header);
+	if (!n64_read_orm_hdr(b, hdr)) {
```
```suggestion
	if (!hdr || !n64_read_orm_hdr(b, hdr)) {
```

ahmed-kamal2004 replied:
did another change like that

wargio commented on test/db/formats/z64:
```diff
@@ -12,14 +12,14 @@ CMDS=iS
      paddr     size      vaddr    vsize align perm name type flags 
 -------------------------------------------------------------------
-0x00001000 0x13f000 0x80001000 0x13f000   0x0 -r-x text      
+0x00001000 0x13f000 0x00100080 0x13f000   0x0 -r-x text      
```
this is wrong, it was correct how it was before.

ahmed-kamal2004 replied:
you are right, fixed it

wargio commented on test/db/formats/z64:
```diff
@@ -12,14 +12,14 @@ CMDS=iS
 CMDS=oml
 EXPECT=<<EOF
- 1 fd: 3 +0x00001000 0x80001000 * 0x8013ffff r-x fmap.text
+ 1 fd: 3 +0x00001000 0x00100080 * 0x0023f07f r-x fmap.text
```
same here. it was correct how it was before.

ahmed-kamal2004 committed:
Solve conversations && Fix issue
